Add email notification for unpaid invoices due soon

nospame opened the pull request:
## Product Description
Adds a new email reminder to billing admins 7 days _before_ an invoice's due date, if the invoice is not yet paid.

## Technical Summary
[SAAS-16796](https://dimagi.atlassian.net/browse/SAAS-16796)
Because similar preexisting emails were predicated on an invoice being _overdue_, I added new methods to get invoices _soon to be_ due, along with the task and email functions themselves. Relates to https://github.com/dimagi/commcare-hq/pull/35932 as it depends on some of the changes there and is intended to be deployed at the same time.

There is some not insignificant refactoring here to reduce duplication of code, as the new invoice reminder is quite similar to the overdue/downgrade notices. Probably best to review by commit with whitespace changes hidden.

## Safety Assurance

### Safety story
There is solid preexisting automated testing here for the downgrade process and associated email notifications. I've also added automated tests for the new invoice reminder emails, which in addition to just checking that the email is sent, also implicitly tests the email context by ensuring it can render (at least the `.txt` template) correctly.


### Automated test coverage
`TestDowngrades` and `TestInvoiceReminder`. Confirmed all tests passed when looking at merging into master (then changed base branch back).

### QA Plan
Not planning QA.

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations
`accounting` migration `0100`, included in this PR, should be rolled back first.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SAAS-16796]: https://dimagi.atlassian.net/browse/SAAS-16796?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nospame committed:
Add INVOICE_REMINDER CommunicationType

nospame committed:
Add html and txt templates for invoice reminder email

nospame committed:
Add tests for InvoiceReminder action

nospame committed:
Add InvoiceReminder action with task

nospame committed:
Add invoicing utils to get invoices due soon

Will be used to send invoice reminder notifications

nospame committed:
Rename downgrade.py => unpaid_invoice.py

nospame committed:
Refactor downgrade_eligible_domains => class-based Downgrade

nospame committed:
Refactor Downgrade to use new UnpaidInvoiceAction base class

nospame committed:
Reorganize Downgrade methods, refactor _apply_downgrade_process

jingcheng16 commented on corehq/apps/accounting/utils/invoicing.py:
```diff
@@ -19,14 +22,33 @@ def _get_all_unpaid_saas_invoices():
+
+def get_oldest_overdue_invoice_over_threshold(today, domain):
+    invoices = _get_unpaid_saas_invoices_in_downgrade_daterange(today)
+    return _get_oldest_invoice_over_threshold(domain, invoices)
```
This is actually `_get_oldest_overdue_invoice_over_threshold`. I wonder does a private function adds value here. The function `get_oldest_overdue_invoice_over_threshold` seems like a wrapper of private function, and it makes the name of both functions hard to differentiate both functions.

jingcheng16 commented on corehq/apps/accounting/mixins.py:
```diff
@@ -8,7 +8,7 @@
 )
 from corehq.apps.accounting.utils.invoicing import (
-    get_oldest_unpaid_invoice_over_threshold,
+    get_oldest_overdue_invoice_over_threshold,
```
Confused about unpaid and overdue, what's the difference between this two words? I saw some places you replaced unpaid with overdue, while some other places you replaced overdue with unpaid

jingcheng16 reviewed (commented):
Only reviewed the first commit. Would appreciate if you could split the refactoring commit into several smaller one, because all those functions you're refactoring have very similar names and it makes it hard to review.

nospame commented on corehq/apps/accounting/utils/invoicing.py:
```diff
@@ -19,14 +22,33 @@ def _get_all_unpaid_saas_invoices():
 
-def get_oldest_unpaid_invoice_over_threshold(today, domain):
-    for overdue_invoice in _get_unpaid_saas_invoices_in_downgrade_daterange(today).filter(
+def _get_domains_over_threshold(invoices, today, get_oldest_invoice_fn):
```
The threshold is the `UNPAID_INVOICE_THRESHOLD`, so in this case $1. For invoices under that amount, we won't bother to return them from these functions.

jingcheng16 replied:
I see. Is it true that we will always ignore invoices under threshold across the codebase? If there are no functions that return invoices under threshold, maybe we don't need to mention "over_threshold" in function names since it is always true? Also can you think of a better name of `UNPAID_INVOICE_THRESHOLD` to indicate it means the amount of money, it is just a nit, I'm not good at naming things, so if you cannot think of any it is fine.

nospame replied:
> Is it true that we will always ignore invoices under threshold across the codebase?

Not quite. There's a different constant `SMALL_INVOICE_THRESHOLD` (set to the same $ amount, but conceptually different) that determines whether or not we set a due date in the first place. In theory these could be different numbers, if say we wanted to set a due date for any invoice over $1, but decide we only want to reach out to customers for unpaid balances over $25 (as an example).

We could call it a "limit" maybe, like `UNPAID_INVOICE_LIMIT`? But I do want to stress that this is a preexisting constant and shouldn't be a focus of review on this PR, if possible, especially since we're aiming to get these changes deployed this week.

nospame commented on corehq/apps/accounting/utils/downgrade.py:
```diff
@@ -24,9 +24,9 @@
-    get_accounts_with_customer_invoices_over_threshold,
-    get_domains_with_subscription_invoices_over_threshold,
-    get_oldest_unpaid_invoice_over_threshold,
+    get_accounts_with_customer_invoices_overdue,
```
This is just a rename for clarity. "Over threshold" is still accurate in that it's over the `UNPAID_INVOICE_THRESHOLD`, but more specifically these are looking for invoices that are past their due date... which is in service of adding additional utils to get invoices that are unpaid, over the `UNPAID_INVOICE_THRESHOLD`, but not due yet.

nospame committed:
Rename 'unpaid' or 'over_threshold' functions to be specifically 'overdue'

nospame committed:
Refactor invoicing utils that get overdue invoices to use more generic helpers

This is to allow the helpers to be used to also get invoices that are unpaid, but not yet due

nospame committed:
Use const for downgrade days overdue

nospame committed:
Reorganize code

jingcheng16 commented on corehq/apps/accounting/utils/invoicing.py:
```diff
@@ -59,7 +60,7 @@ def _get_domains_over_threshold(invoices, today, get_oldest_invoice_fn):
 
 
 def get_accounts_with_customer_invoices_overdue(today):
```
Want to confirm my understanding of the operation here. All accounts returned by this function won't be downgraded yet, right?

nospame replied:
That's right. There's still the check to make sure we don't automatically downgrade Customer Billing Accounts, but they will get the invoice reminder, the overdue notice, and the downgrade _warning_ emails in the normal timelines.

https://github.com/dimagi/commcare-hq/blob/fbef7976c6c4b6af71e00d6fcbe27454b8346f82/corehq/apps/accounting/utils/unpaid_invoice.py#L207-L210

jingcheng16 reviewed (commented):
Looks good to me. Just some questions.

nospame commented on corehq/apps/accounting/utils/invoicing.py:
```diff
@@ -40,38 +45,48 @@ def get_oldest_overdue_invoice_over_threshold(today, domain):
-    overdue_customer_invoices_in_downgrade_daterange = unpaid_customer_invoices.filter(
-        date_due__lte=today - datetime.timedelta(days=1),
-        date_due__gte=today - datetime.timedelta(days=61)
+    unpaid_customer_invoices_in_daterange = unpaid_customer_invoices.objects.filter(
```
Not necessarily. While these invoices are all unpaid, `get_accounts_with_customer_invoices_due_soon` will get the ones that are due within a week from the current date - those ones aren't overdue yet. 

This is a good thought though, and made me realize `_get_oldest_invoice_over_threshold` is still using the wrong variable name (`overdue_invoice` instead of `unpaid_invoice`).

nospame committed:
Improve variable naming

nospame committed:
Correct a couple more 'overdue' to 'unpaid' variable names

jingcheng16 commented:
@nospame Thanks for addressing the feedbacks! Do you know why this PR don't have test running? Is it because the base branch is not master?

biyeun commented on corehq/apps/accounting/utils/invoicing.py:
```diff
@@ -40,38 +45,48 @@ def get_oldest_overdue_invoice_over_threshold(today, domain):
+    return _get_accounts_over_threshold(date_start, date_end)
+
+
+def _get_accounts_over_threshold(date_start, date_end):
```
naming seems confusing because it yields invoice, total...not account

biyeun replied:
`unpaid_customer_invoices_over_threshold`

biyeun commented on corehq/apps/accounting/utils/invoicing.py:
```diff
@@ -40,38 +45,48 @@ def get_oldest_overdue_invoice_over_threshold(today, domain):
+    return _get_domains_over_threshold(invoices, today, get_oldest_overdue_invoice_over_threshold)
+
+
+def _get_domains_over_threshold(invoices, today, get_oldest_invoice_fn):
```
similarly confusing name to the `_get_accounts_over_threshold` based on what it yields.

maybe for this function, `_get_overdue_invoice_criteria_for_domains`?

it might be helpful to add doc strings about what the expected value is and what the params are, and what the function's purpose is

biyeun commented on corehq/apps/accounting/templates/accounting/email/invoice_reminder.html:
```diff
@@ -0,0 +1,94 @@
+</p>
+
+{% blocktrans %}
+  <p>You can pay this bill by credit card or by wire:</p>
```
wrap just the text in each block element with `blocktrans`

biyeun commented on corehq/apps/accounting/templates/accounting/email/invoice_reminder.html:
```diff
@@ -0,0 +1,94 @@
+
+<p>
+  {% blocktrans %}
+    Best Regards,<br />
```
inline trans to avoid the `<br />`

biyeun replied:
inline trans to avoid the `<br />`

biyeun commented on corehq/apps/accounting/utils/downgrade.py:
```diff
@@ -70,6 +70,40 @@ def run_action(cls, only_downgrade_domain=None):
                 )
 
+    @classmethod
+    def _apply_downgrade_process(cls, oldest_unpaid_invoice, total, today, subscription):
```
could `Downgrade` be turned into an instance where a lot of these arguments are passed in initially?

biyeun commented on corehq/apps/accounting/utils/downgrade.py:
```diff
@@ -101,6 +105,80 @@ def _get_communication_model_context(domain, oldest_unpaid_invoice):
+            bcc = [settings.GROWTH_EMAIL]
+
+        subject = _(
+            "Your CommCare Billing Statement for {} is due in {} days".format(
```
translation should wrap inside `format`

biyeun commented on corehq/apps/accounting/utils/downgrade.py:
```diff
@@ -101,6 +105,80 @@ def _get_communication_model_context(domain, oldest_unpaid_invoice):
+            cc=[settings.ACCOUNTS_EMAIL],
+            bcc=bcc,
+            email_from=get_dimagi_from_email())
+        communication_model.objects.create(
```
do we want to do this as a callback function in the task? will that work, is it picklable? or do we create a separate accounting tasks that sends the email synchronously and then when it is done, create the communication model

biyeun replied:
might be a good idea to also go through the old code and update it where we do this incorrectly

nospame replied:
There is a really simple solution here which is just to replace the use of `corehq.apps.hqwebapp.tasks.send_html_email_async` in this file with `corehq.util.log.send_HTML_email`, which is what the `_async` function itself calls. What would you think of that?

biyeun replied:
yes, i think that makes sense. how would we handle failures and retries?

nospame replied:
If something fails while sending the email it should raise an exception, which since it's synchronous it will bubble up to the try/catch in `run_action`, to get logged as a failure to downgrade (or send reminder) for this project space or billing account:
https://github.com/dimagi/commcare-hq/blob/47dd3ab285e421c865cc17487dfc4935d0db4b79/corehq/apps/accounting/utils/unpaid_invoice.py#L49-L57

I can't tell how necessary a retry mechanism would be (I'm not clear if that's typically handling a failure of the task related to celery, or a failure of the email to send for other reasons). I do see that in other places we use `send_HTML_email` we don't add retries, which makes me think that it should be fine.

But if you'd be more comfortable seeing this as a separate accounting task that explicitly links the email sending (with retries) to creating the `CommunicationHistory`, I can do that instead.

biyeun replied:
just because we do it poorly elsewhere doesn't mean it's a good idea. I think let's use `send_HTML_email` for now to speed up dev, and split it into a task as a followup PR. I think we should at least escape errors and log them in order to make sure the entire task for all of the downgrade/reminder actions doesn't fail as a result of an email error. A separate task is nice so that the email goes to the proper celery queue for emails and it doesn't block this task as sometimes that can be a time consuming action

nospame replied:
Okay, here's the ticket I created for follow-ups, including this task. https://dimagi.atlassian.net/browse/SAAS-16820

biyeun reviewed (commented):
will give this a ✅ once translation comments and the note about synchronous email and communication model gets addressed. the refactor comments / naming can be addressed later.

nospame committed:
Improve invoice reminder translations

nospame committed:
Remove bad use of .objects

nospame committed:
Send unpaid invoice emails synchronously

nospame committed:
Merge branch 'master' into em/invoice-due-soon-reminder

mjriley reviewed (approved):
Changes look good, thanks!
